File: src/codes/tight-coupling/py/rfc_mediator.py
from abc import ABC, abstractmethod

class Mediator(ABC):
    @abstractmethod
    def notify(self, sender: str, data: any):
        pass

class Mailer:
    def send_welcome(self, email: str):
        print(f"送信しました: {email} → ようこそ！")

class AppMediator(Mediator):
    def __init__(self, mailer: Mailer):
        self.mailer = mailer

    def notify(self, sender: str, data: any):
        if sender == "UserService":
            self.mailer.send_welcome(data)

class UserService:
    def __init__(self, mediator: Mediator):
        self.mediator = mediator

    def register_user(self, email: str):
        print(f"ユーザー登録: {email}")
        self.mediator.notify("UserService", email)

# 利用例
mailer = Mailer()
mediator = AppMediator(mailer)
user_service = UserService(mediator)
user_service.register_user("user@example.com")

File: src/codes/tight-coupling/ts/rfc_mediator.ts
// Mediator インターフェース
interface Mediator {
  notify(sender: string, data: any): void;
}

// コンポーネント（UserService）
class UserService {
  constructor(private mediator: Mediator) {}

  registerUser(email: string): void {
    console.log(`ユーザー登録: ${email}`);
    this.mediator.notify("UserService", email);
  }
}

// コンポーネント（Mailer）
class Mailer {
  sendWelcome(email: string): void {
    console.log(`送信しました: ${email} → ようこそ！`);
  }
}

// Mediator 実装
class AppMediator implements Mediator {
  constructor(private mailer: Mailer) {}

  notify(sender: string, data: any): void {
    if (sender === "UserService") {
      this.mailer.sendWelcome(data);
    }
  }
}

// 利用例
const mailer = new Mailer();
const mediator = new AppMediator(mailer);
const userService = new UserService(mediator);

userService.registerUser("user@example.com");
